tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs

```diff
@@ -0,0 +1,424 @@
+        X509Certificate? certificate,
+        X509Chain? chain,
+        SslPolicyErrors sslPolicyErrors)
+    {
+        if (sslPolicyErrors == SslPolicyErrors.None)
+            return true;
+
+        // Return true if client certificate is not required
+        if (sslPolicyErrors == SslPolicyErrors.RemoteCertificateNotAvailable)
+            return true;
+        
+        // Validate client certificate with a custom chain
+        if (certificate is not null)
+        {
+            var isValid = RootChain.Build(new X509Certificate2(certificate));
+            if (!isValid)
+            {
+                foreach (var status in RootChain.ChainStatus)
+                {
+                    Log.Error("[Server] Chain error: {0}", status.StatusInformation);
+                }
+            }
+
+            return isValid;
+        }
+
+        // Refuse everything else
+        Log.Error("[Server] Certificate error: {0}", sslPolicyErrors);
+        return false;
+    }    
+
+    private bool ValidateServerCertificate(
+        object sender,
+        X509Certificate? certificate,
+        X509Chain? chain,
+        SslPolicyErrors errors)
+    {
+        if (errors == SslPolicyErrors.None)
+            return true;
+
+        // Missing cert or the destination hostname wasn't valid for the cert.
+        if ((errors & ~SslPolicyErrors.RemoteCertificateChainErrors) != 0)
+            return false;
+
+        // Validate client certificate with a custom chain
+        if (certificate is not null)
+        {
+            chain ??= RootChain;
+            var isValid = chain.Build(new X509Certificate2(certificate));
+            if (!isValid)
+            {
+                foreach (var status in chain.ChainStatus)
+                {
+                    Log.Error("[Client] Chain error: [{0}] {1}", status.Status, status.StatusInformation);
+                }
+            }
+
+            return isValid;
+        }
+        
+        // Refuse everything else
+        Log.Error("[Client] Certificate error: {0}", errors);
+        return false;
+    }
+    
+    public override async Task<IMqttClient> CreateClient()
+    {
+        var client = await ClientFactory.CreateTcpClient(DefaultConnectOptions, DefaultTcpOptions);
+        return client;
+    }
+
+    private MqttClientTcpOptions DefaultTcpOptions => new("localhost", 21883)
+    {
+        TlsOptions = new ClientTlsOptions
+        {
+            UseTls = true,
+            SslOptions = new SslClientAuthenticationOptions
+            {
+                TargetHost = "localhost",
+                CertificateChainPolicy = RootChainPolicy
+            }, 
+            ServerCertificateValidator = ValidateServerCertificate
+        }
+    };
+
+    protected override void AfterAll()
+    {
+        // shut down our local TCP server
+        _server.Shutdown();
+        base.AfterAll();
+    }
+
+    [Fact]
+    public async Task ShouldAutomaticallyReconnectAndSubscribeAfterServerDisconnect()
```

Comment:
Would it be better if we just made the TCP end 2 end specs sub-classable and reuse those @Arkatufus ?

Reply:
I can give that a try

Reply:
Done